pkg/cmd/release/github.go

```diff
@@ -0,0 +1,163 @@
+	// really gets from that email is "there's a LOT of blockers to resolve".
+	openIssueMaxPages = 3
+)
+
+type githubClient interface {
+	openIssues(labels []string) ([]githubIssue, error)
+}
+
+// githubClientImpl implements the githubClient interface.
+type githubClientImpl struct {
+	client *github.Client
+	ctx    context.Context
+	owner  string
+	repo   string
+}
+
+var _ githubClient = &githubClientImpl{}
+
+type githubIssue struct {
+	Number      int
+	ProjectName string
+}
+
+// newGithubClient returns a githubClient, which is wrapper for a *github.Client
+// for a specific repository.
+// To generate personal access token, go to https://github.com/settings/tokens.
+func newGithubClient(ctx context.Context, accessToken string) *githubClientImpl {
+	ts := oauth2.StaticTokenSource(
+		&oauth2.Token{AccessToken: accessToken},
+	)
+	tc := oauth2.NewClient(ctx, ts)
+	return &githubClientImpl{
+		client: github.NewClient(tc),
+		ctx:    ctx,
+		owner:  "cockroachdb",
+		repo:   "cockroach",
+	}
+}
+
+func (c *githubClientImpl) openIssues(labels []string) ([]githubIssue, error) {
+	var details []githubIssue
+	for pageNum := 1; pageNum <= openIssueMaxPages; pageNum++ {
+		issues, _, err := c.client.Issues.List(
+			c.ctx,
+			true, /* all: true searches `/issues/`, false searches `/user/issues/` */
+			&github.IssueListOptions{
+				Filter: "all",
+				State:  "open",
+				Labels: labels,
+				ListOptions: github.ListOptions{
+					PerPage: githubPageSize,
+					Page:    pageNum,
+				},
+			},
+		)
+		if err != nil {
+			return nil, err
+		}
+		for _, i := range issues {
+			issueNum := *i.Number
+			projectName, err := c.mostRecentProjectName(issueNum)
+			if err != nil {
+				return nil, err
+			}
+
+			details = append(details, githubIssue{
+				Number:      issueNum,
+				ProjectName: projectName,
+			})
+		}
+
+		if len(issues) < githubPageSize {
+			break
+		}
+	}
+
+	return details, nil
+}
+
+func (c *githubClientImpl) branchExists(branchName string) (bool, error) {
+	protected := true
+	branches, _, err := c.client.Repositories.ListBranches(
+		c.ctx, c.owner, c.repo,
+		&github.BranchListOptions{Protected: &protected},
+	)
+	if err != nil {
+		return false, err
+	}
+	for _, b := range branches {
+		if *b.Name == branchName {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// mostRecentProjectName returns the most recently added project name, by iterating
+// through issue.Project.Name values found in the issues' timeline.
+// Returns nil if no project name found.
+func (c *githubClientImpl) mostRecentProjectName(issueNum int) (string, error) {
+	pageNum := 1
+	removedProjects := make(map[string]bool)
+	for {
+		events, _, err := c.client.Issues.ListIssueTimeline(
+			c.ctx, c.owner, c.repo, issueNum, &github.ListOptions{
+				PerPage: githubPageSize,
+				Page:    pageNum,
+			})
+		if err != nil {
+			return "", err
+		}
+		for i := len(events) - 1; i >= 0; i-- {
```

Comment:
TODO - double-check this API:

I think this needs to start at i = 0 and increment by +1. Looking at this example, event[0] is the most recent, which is where I want to start:

https://api.github.com/repos/cockroachdb/cockroach/issues/66606/timeline